File: processor/src/main/java/io/shreyash/rush/ExtensionProcessor.java
package io.shreyash.rush;

import com.google.appinventor.components.annotations.SimpleEvent;
import com.google.appinventor.components.annotations.SimpleFunction;
import com.google.appinventor.components.annotations.SimpleProperty;
import com.google.auto.service.AutoService;

import org.xml.sax.SAXException;

import java.io.IOException;
import java.lang.annotation.Annotation;
import java.util.Set;

import javax.annotation.processing.AbstractProcessor;
import javax.annotation.processing.Messager;
import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.Processor;
import javax.annotation.processing.RoundEnvironment;
import javax.annotation.processing.SupportedAnnotationTypes;
import javax.annotation.processing.SupportedSourceVersion;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.Element;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.tools.Diagnostic;
import javax.xml.parsers.ParserConfigurationException;

import io.shreyash.rush.blocks.BlockStore;
import io.shreyash.rush.blocks.DesignerProperty;
import io.shreyash.rush.blocks.Event;
import io.shreyash.rush.blocks.Method;
import io.shreyash.rush.blocks.Property;

@AutoService(Processor.class)
@SupportedSourceVersion(SourceVersion.RELEASE_8)
@SupportedAnnotationTypes({
    "com.google.appinventor.components.annotations.SimpleEvent",
    "com.google.appinventor.components.annotations.SimpleFunction",
    "com.google.appinventor.components.annotations.SimpleProperty",
    "com.google.appinventor.components.annotations.DesignerProperty"
})
public class ExtensionProcessor extends AbstractProcessor {
  private final BlockStore store = BlockStore.getInstance();

  private boolean isFirstRound = true;
  private Messager messager;

  @Override
  public synchronized void init(ProcessingEnvironment processingEnv) {
    super.init(processingEnv);
    this.messager = processingEnv.getMessager();
  }

  @Override
  public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
    if (!this.isFirstRound || this.processingEnv.getOptions().get("root") == null) {
      return true;
    }
    this.isFirstRound = false;

    final String org = this.processingEnv.getOptions().get("org");
    final String extName = this.processingEnv.getOptions().get("extName");

    // Process all SimpleEvents
    for (Element el : roundEnv.getElementsAnnotatedWith(SimpleEvent.class)) {
      if (!isInRightParent(el, extName, org, "@SimpleEvent")) {
        continue;
      }

      if (isPublic(el, "@SimpleEvent")) {
        Event event = new Event(el, this.messager);
        this.store.putEvent(event);
      }
    }

    // Process all SimpleFunctions
    for (Element el : roundEnv.getElementsAnnotatedWith(SimpleFunction.class)) {
      if (!isInRightParent(el, extName, org, "@SimpleFunction")) {
        continue;
      }

      if (isPublic(el, "@SimpleFunction")) {
        Method method = new Method(el, this.messager);
        this.store.putMethod(method);
      }
    }

    // Process all SimpleProps
    for (Element el : roundEnv.getElementsAnnotatedWith(SimpleProperty.class)) {
      if (!isInRightParent(el, extName, org, "@SimpleProperty")) {
        continue;
      }

      if (isPublic(el, "@SimpleProperty")) {
        Property prop = new Property(el, this.messager);
        this.store.putProperty(prop);
      }
    }

    // Process all DesignerProps
    for (Element el : roundEnv.getElementsAnnotatedWith(com.google.appinventor.components.annotations.DesignerProperty.class)) {
      if (!isInRightParent(el, extName, org, "@DesignerProperty")) {
        continue;
      }

      if (isPublic(el, "@DesignerProperty")) {
        DesignerProperty prop = new DesignerProperty(el, this.messager);
        this.store.putDesignerProperty(prop);
      }
    }

    generateInfoFiles(extName, org);

    return false;
  }

  private void generateInfoFiles(String extName, String org) {
    String root = this.processingEnv.getOptions().get("root");
    String version = this.processingEnv.getOptions().get("version");
    String type = org + "." + extName;
    String output = this.processingEnv.getOptions().get("output");

    InfoFilesGenerator generator = new InfoFilesGenerator(root, version, type, output);
    try {
      generator.generateComponentsJson();
      generator.generateBuildInfoJson();
    } catch (IOException | ParserConfigurationException | SAXException e) {
      this.messager.printMessage(Diagnostic.Kind.ERROR, e.getMessage());
    }
  }

  private boolean isInRightParent(Element el, String extName, String org, String annotationName) {
    final boolean res = el.getEnclosingElement().getSimpleName().toString().equals(extName);

    if (!res) {
      this.messager.printMessage(Diagnostic.Kind.ERROR,
          "Annotation" + annotationName + "can't be used on element \"" + el.getSimpleName()
              + "\". It can only be used on members of class \"" + org + "." + extName + "\".");
    }

    return res;
  }

  private boolean isPublic(Element el, String annotationName) {
    final boolean isPublic = el.getModifiers().contains(Modifier.PUBLIC);

    if (!isPublic) {
      messager.printMessage(Diagnostic.Kind.ERROR,
          "Private element \"" + el.getSimpleName() + "\" can't be annotated with" + annotationName + " .");
    }

    return isPublic;
  }
}

File: processor/src/main/java/io/shreyash/rush/processor/ExtensionProcessor.kt
package io.shreyash.rush.processor

import com.google.appinventor.components.annotations.Meta
import com.google.appinventor.components.annotations.SimpleEvent
import com.google.appinventor.components.annotations.SimpleFunction
import com.google.appinventor.components.annotations.SimpleProperty
import com.google.auto.service.AutoService
import io.shreyash.rush.processor.block.DesignerProperty
import io.shreyash.rush.processor.block.Event
import io.shreyash.rush.processor.block.Function
import io.shreyash.rush.processor.block.Property
import io.shreyash.rush.processor.model.Extension
import javax.annotation.processing.*
import javax.lang.model.SourceVersion
import javax.lang.model.element.Element
import javax.lang.model.element.Modifier
import javax.lang.model.element.TypeElement
import javax.lang.model.util.Elements
import javax.tools.Diagnostic

@AutoService(Processor::class)
@SupportedSourceVersion(SourceVersion.RELEASE_8)
@SupportedAnnotationTypes(
    "com.google.appinventor.components.annotations.Meta",
    "com.google.appinventor.components.annotations.SimpleEvent",
    "com.google.appinventor.components.annotations.SimpleFunction",
    "com.google.appinventor.components.annotations.SimpleProperty",
    "com.google.appinventor.components.annotations.DesignerProperty"
)
class ExtensionProcessor : AbstractProcessor() {
    private var isFirstRound = true

    private lateinit var messager: Messager
    private lateinit var elementUtil: Elements

    @Synchronized
    override fun init(processingEnv: ProcessingEnvironment) {
        super.init(processingEnv)
        messager = processingEnv.messager
        elementUtil = processingEnv.elementUtils
    }

    override fun process(annotations: Set<TypeElement?>, roundEnv: RoundEnvironment): Boolean {
        if (!this.isFirstRound) {
            return true
        }
        this.isFirstRound = false

        val extensions = roundEnv.getElementsAnnotatedWith(Meta::class.java).map {
            processExtensionElement(it)
        }
        generateInfoFiles(extensions)

        return false
    }

    private fun processExtensionElement(element: Element): Extension {
        // Process simple events
        val events = element.enclosedElements
            .filter {
                it.getAnnotation(SimpleEvent::class.java) != null && isPublic(it)
            }.map { Event(it, this.messager) }

        // Process simple functions
        val functions = element.enclosedElements
            .filter {
                it.getAnnotation(SimpleFunction::class.java) != null && isPublic(it)
            }.map { Function(it, messager) }

        // Process simple properties
        val processedProperties = mutableListOf<Property>()
        val properties = element.enclosedElements
            .filter {
                it.getAnnotation(SimpleProperty::class.java) != null && isPublic(it)
            }.map {
                val property = Property(it, this.messager, processedProperties)
                processedProperties.add(property)
                property
            }

        // Process designer properties
        val designerProperties = element.enclosedElements
            .filter {
                it.getAnnotation(com.google.appinventor.components.annotations.DesignerProperty::class.java) != null
                        && isPublic(it)
            }.map {
                DesignerProperty(it, this.messager, properties)
            }

        val packageName = this.elementUtil.getPackageOf(element).qualifiedName.toString()
        val fqcn = "$packageName.${element.simpleName}"

        return Extension(
            element.getAnnotation(Meta::class.java),
            fqcn,
            events,
            functions,
            properties,
            designerProperties
        )
    }

    /** Generates the component info files (JSON). */
    private fun generateInfoFiles(extensions: List<Extension>) {
        val projectRoot = processingEnv.options["projectRoot"]!!
        val outputDir = processingEnv.options["outputDir"]!!

        val generator = InfoFilesGenerator(projectRoot, outputDir, extensions)
        try {
            generator.generateComponentsJson()
            generator.generateBuildInfoJson()
        } catch (e: Throwable) {
            messager.printMessage(Diagnostic.Kind.ERROR, e.message ?: e.stackTraceToString())
        }
    }

    /** @returns `true` if [element] is a public element. */
    private fun isPublic(element: Element): Boolean {
        val isPublic = element.modifiers.contains(Modifier.PUBLIC)
        if (!isPublic) {
            messager.printMessage(
                Diagnostic.Kind.WARNING,
                "Element \"${element.simpleName}\" is private. It should be public to be visible in" +
                        " the blocks editor."
            )
        }
        return isPublic
    }
}
